File: back/search/src/repository/models/recipe.rs
use bson::Bson;
use bson::oid::ObjectId;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Recipe {
    #[serde(with = "bson::serde_helpers::chrono_datetime_as_bson_datetime")]
    pub updated_at: DateTime<Utc>,
    pub author_id: ObjectId,
    pub title: String,
    pub sum_rating: u32,
    pub count_rating: u32,
    pub description: String,
    pub prep_time: u32,
    pub steps: Vec<String>,
    pub ingredients: Vec<String>,
    pub allergens: Vec<String>,
    pub tags: Vec<String>,
    pub tokens: Vec<String>,
    pub ratings: Vec<ObjectId>,
}

impl Recipe {
    #[must_use]
    pub const fn get_collection_name() -> &'static str {
        "report"
    }
}

File: back/generic/src/main/java/app/recipe/model/Recipe.java
package app.recipe.model;

import app.allergen.model.Allergen;
import app.rating.model.Rating;
import app.tag.model.Tag;
import app.user.model.User;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.bson.types.ObjectId;

import java.util.*;

@Getter
@Setter
@NoArgsConstructor
public class Recipe {
    // Nu se schimba ever
    private ObjectId recipeId;
    private User authorId;

    private Date updatedAt;

    private String title;
    private List<String> ingredients;
    private String description;
    private List<String> steps;
    private int prepTime; // in minute
    private Set<Tag> tags;
    private List<String> tokens; // date de AI
    private List<Allergen> allergens;


    private List<Rating> ratings;
    private int sumRating;
    private int countRating;

    public Recipe(User authorId, String title, List<String> ingredients, String desc, List<String> steps, int prepTime, List<Allergen> allergens){
        this.authorId = authorId;
        this.title = title;
        this.ingredients = ingredients;
        this.description = desc;
        this.steps = steps;
        this.prepTime = prepTime;
        this.allergens = allergens;
        this.tags = new HashSet<>();
        this.tokens = new ArrayList<>();
        this.ratings = new ArrayList<>();
        this.sumRating = 0;
        this.countRating = 0;
        this.updatedAt = new Date();
    }

    public void modifyRecipe(Recipe newRecipe) {
        this.title = newRecipe.getTitle();
        this.ingredients = newRecipe.getIngredients();
        this.description = newRecipe.getDescription();
        this.steps = newRecipe.getSteps();
        this.prepTime = newRecipe.getPrepTime();
        this.allergens = newRecipe.getAllergens();
        this.updatedAt = new Date();
    }

    public void addRating(Rating rating) {
        this.ratings.add(rating);
        this.sumRating += (int) rating.getValue();
        this.countRating++;
    }
    public void addTag(Tag tag) {
        tags.add(tag);
    }
    public void addToken(String token) {
        tokens.add(token);
    }
    public float getRating() {
        if (countRating == 0) {
            return 0;
        }
        return (float) sumRating / countRating;
    }
}
